src/main/java/com/mraof/minestuck/entity/AttackingAnimatedEntity.java

```diff
@@ -0,0 +1,116 @@
+package com.mraof.minestuck.entity;
+
+import net.minecraft.entity.EntityType;
+import net.minecraft.entity.LivingEntity;
+import net.minecraft.entity.ai.goal.MeleeAttackGoal;
+import net.minecraft.world.World;
+
+/**
+ * A base class for animated entities with a potentially delayed attack.
+ */
+public abstract class AttackingAnimatedEntity extends AnimatedCreatureEntity
+{
+	/**
+	 * The delay between the start of the animation and the moment the damage lands
+	 */
+	protected int attackDelay;
+	
+	/**
+	 * The delay after an attack and before another start
+	 */
+	protected int attackRecovery;
+	
+	protected AttackingAnimatedEntity(EntityType<? extends AttackingAnimatedEntity> type, World world)
+	{
+		super(type, world);
+	}
+	
+	@Override
+	protected void registerGoals()
+	{
+		super.registerGoals();
+		goalSelector.addGoal(3, new AttackingAnimatedEntity.DelayedAttackGoal(this, 1F, false));
+	}
+	
+	@Override
+	protected void endTimedAction(Actions action)
+	{
+		if(action == Actions.ATTACK)
+		{
+			this.setCurrentAction(Actions.ATTACK_RECOVERY, attackRecovery);
+			performAttack();
+		} else
+			super.endTimedAction(action);
+	}
+	
+	private void performAttack()
+	{
+		this.onAttackEnd();
+		
+		if(getTarget() != null && isInRange(getTarget()))
+		{
+			doHurtTarget(getTarget());
+			// TODO: AOE bounding box collision checks + aoe flag
+		}
+	}
+	
+	private boolean isInRange(LivingEntity target)
+	{
+		double reach = this.getBbWidth() * 2.0F * this.getBbWidth() * 2.0F + target.getBbWidth();
+		return this.distanceToSqr(target) <= reach;
+	}
+	
+	/**
+	 * Starts a long attack against the current target if this entity isn't already performing some action with a duration.
+	 * Useful to sync animations and add extra delays
+	 */
+	private void startAttack()
+	{
+		if(!this.hasTimedAction())
+		{
+			this.setCurrentAction(Actions.ATTACK, attackDelay);
+			
+			this.onAttackStart();
+		}
+	}
+	
+	/**
+	 * Is called when an attack starts. Can be extended to apply effects during an attack.
+	 */
+	protected void onAttackStart()
+	{
+	}
+	
+	/**
+	 * Is called when an attack starts. Can be extended to remove effects once an attack has ended.
+	 */
+	protected void onAttackEnd()
+	{
+	}
```

Comment:
Can you foresee any scenario in which modifiers set during these functions could accidentally get applied multiple times without having been removed? For example if the game is quit during an attack

Reply:
For your example, it depends on the type of modifier. If it's a "permanent" modifier, it will be serialized, and thus will be remain when the entity is unloaded and reloaded. If it's a transient modifier, it will not be serialized', and thus be gone when the entity is unloaded and reloaded.

However, a possible scenario is that a delayed attack starts, and something else calls `setAction()` and changes the action before it has completed. There is no such call anywhere right now, but this shows a problem with the current action system, where actions can technically be interrupted, but the interruption is not handled.

One option is to strictly disallow action interruption in `setAction()`, or handle interruption such that the modifier gets properly removed. I'm leaning towards a third alternative that prevents outside interruptions by moving the attack timer into the ai goal. Doing so would also mean that the action state for attacking entites could be simplified to a "isAttacking" property.

Reply:
Do you believe it is likely to be a problem or something we could hold off on?

Reply:
As said, it's not a problem right now, but it might become a problem if we start getting more types of actions introduced. I think we can hold off on this together with my ideas from https://github.com/mraof/Minestuck/pull/346#issuecomment-1146206387